python/ray/tune/trial_runner.py

```diff
@@ -271,34 +274,90 @@ def checkpoint(self, force=False):
         """
         if not self._local_checkpoint_dir:
             return
+
+        if self._checkpoint_thread and force:
+            # Checkpoint forced, but a thread is already running.
+            # In this case, wait until the checkpoint finished and
+            # then directly schedule a new one.
+            self._checkpoint_thread.join()
+
+        if self._checkpoint_thread and \
+           not self._checkpoint_thread.is_alive() and \
+           not self._checkpoint_finished.is_set():
+
+            raise RuntimeError(
+                "Checkpoint thread finished but the stop event was not "
+                "set. This indicates that the checkpointing operation "
+                "failed. Please check the logs to find out what went wrong.")
+
+        elif self._checkpoint_finished.is_set():
+            self._checkpoint_thread.join()
+            self._checkpoint_thread = None
+            self._checkpoint_finished.clear()
+
+            self._syncer.sync_up_if_needed()
```

Comment:
is this needed here?

Reply:
We could also move it to the checkpointing thread. Actually we might want to do that (as syncing to cloud might be time consuming)